pkg/controller/v1alpha1/inferencegraph/controller.go

```diff
@@ -173,9 +178,53 @@ func (r *InferenceGraphReconciler) Reconcile(ctx context.Context, req ctrl.Reque
 		return reconcile.Result{}, errors.Wrapf(err, "fails to create DeployConfig")
 	}
 
+	// name of our custom finalizer
+	finalizerName := "inferencegraph.finalizers"
+
+	// examine DeletionTimestamp to determine if object is under deletion
+	if graph.ObjectMeta.DeletionTimestamp.IsZero() {
+		// The object is not being deleted, so if it does not have our finalizer,
+		// then lets add the finalizer.
+		if !utils.Includes(graph.ObjectMeta.Finalizers, finalizerName) {
+			graph.ObjectMeta.Finalizers = append(graph.ObjectMeta.Finalizers, finalizerName)
+			patchYaml := "metadata:\n  finalizers: [" + strings.Join(graph.ObjectMeta.Finalizers, ",") + "]"
+			patchJson, _ := yaml.YAMLToJSON([]byte(patchYaml))
+			if err = r.Patch(ctx, graph, client.RawPatch(types.MergePatchType, patchJson)); err != nil {
+				return reconcile.Result{}, err
+			}
+		}
+	} else {
+		// The object is being deleted
+		if utils.Includes(graph.ObjectMeta.Finalizers, finalizerName) {
+			// our finalizer is present, so lets handle any external dependency
+			if err = r.onDeleteCleanup(ctx, graph); err != nil {
+				// if fail to delete the external dependency here, return with error
+				// so that it can be retried
+				return ctrl.Result{}, err
+			}
+
+			// remove our finalizer from the list and update it.
+			graph.ObjectMeta.Finalizers = utils.RemoveString(graph.ObjectMeta.Finalizers, finalizerName)
+			patchYaml := "metadata:\n  finalizers: [" + strings.Join(graph.ObjectMeta.Finalizers, ",") + "]"
+			patchJson, _ := yaml.YAMLToJSON([]byte(patchYaml))
+			if err = r.Patch(ctx, graph, client.RawPatch(types.MergePatchType, patchJson)); err != nil {
+				return reconcile.Result{}, err
+			}
+		}
+
+		// Stop reconciliation as the item is being deleted
+		return ctrl.Result{}, nil
+	}
+
 	deploymentMode := isvcutils.GetDeploymentMode(graph.Status.DeploymentMode, graph.ObjectMeta.Annotations, deployConfig)
 	r.Log.Info("Inference graph deployment ", "deployment mode ", deploymentMode)
 	if deploymentMode == constants.RawDeployment {
+		// If the inference graph has auth enabled, create the supporting resources
+		err = handleInferenceGraphRawAuthResources(ctx, r.Clientset, r.Scheme, graph)
+		if err != nil {
+			return ctrl.Result{}, errors.Wrapf(err, "fails to reconcile resources for auth verification")
+		}
+
 		// Create inference graph resources such as deployment, service, hpa in raw deployment mode
 		deployment, url, err := handleInferenceGraphRawDeployment(r.Client, r.Clientset, r.Scheme, graph, routerConfig)
```

Comment:
Looking at the controller flow from here, `handleInferenceGraphRawDeployment` -> `NewRawKubeReconciler` -> `NewDeploymentReconciler` -> `createRawDeploymentODH` where we have 
```
if val, ok := componentMeta.Labels[constants.ODHKserveRawAuth]; ok && val == "true" {
			err := addOauthContainerToDeployment(clientset, deployment, componentMeta, componentExt, podSpec)
			if err != nil {
				return nil, err
			}
		}
```
Does this mean that if the graph has the `ODHKserveRawAuth` label, the graph deployment will get the `oauth-proxy` container. Not sure if I'm missing something 🤔

Reply:
In my testings it didn't get oauth-proxy. It could be because of [RHOAIENG-20326](https://issues.redhat.com/browse/RHOAIENG-20326) which we discussed today.

Most likely, when RHOAIENG-20326 is fixed, the IG will get the proxy. If that's the case, I'll appreciate you create a follow-up ticket for fixing it (although you'd be kind if you fix it as part of RHOAIENG-20326 :smile: ).

Reply:
To me this feels like more in scope with this PR whereas https://issues.redhat.com/browse/RHOAIENG-20326 should IMO be limited to making the switch from label to annotation

Reply:
It does not fit here, nor in [RHOAIENG-20326](https://issues.redhat.com/browse/RHOAIENG-20326). With the current code, there is nothing to fix in this PR. This is my motivation for that to be a follow-up Jira.

By doing any fix here, I'd need to ensure no regression, and also, I'd need to also implement RHOAIENG-20326, just to ensure that the fix will actually work.

Reply:
Issue created: https://issues.redhat.com/browse/RHOAIENG-21300